Implemented Host that includes a routing system.

aratz-lasa opened the pull request:
I want to resolve #280 explicitly distinguishing different Host types (Basic, Routed, Relayed) like in Go's implementation. 
I have implemented RoutedHost in a separate module from BasicHost. An alternative would be to rename [basic_host](https://github.com/libp2p/py-libp2p/blob/master/libp2p/host/basic_host.py) module to _**host**_, and include all three types in that same module.

mhchia commented:
Nice! The structure in your PR looks good to me. Would you mind adding some tests for the routed host? It will be great to have a case which successfully find the address using routing and another one failing to do that.

aratz-lasa commented:
I must clarify some things:

- In order to carry out [RoutedHost tests](https://github.com/aratz-lasa/py-libp2p/blob/issue-280/tests/host/test_routed_host.py), I had to be able to save a TCP addrs for a given Host in the routing DHT. However, [find_peer](https://github.com/libp2p/py-libp2p/blob/00f83a36948fac766be2e2319d59fad86387a87c/libp2p/routing/kademlia/kademlia_peer_router.py#L17) current implementation uses [decode_peerinfo](https://github.com/libp2p/py-libp2p/blob/00f83a36948fac766be2e2319d59fad86387a87c/libp2p/routing/kademlia/kademlia_peer_router.py#L30), and therefore, [create_kad_peerinfo](https://github.com/libp2p/py-libp2p/blob/00f83a36948fac766be2e2319d59fad86387a87c/libp2p/kademlia/kad_peerinfo.py#L141) is used. To sum up, currently just UDP addrs can be estored in Kademlia because **create_kad_peerinfo** loads every peerinfo as if it was a UDP address. So, I have changed **find_peer** to use a different decoder [PeerInfo.info_from_string](https://github.com/aratz-lasa/py-libp2p/blob/814356383160fb30d376850710d240d9360cea91/libp2p/routing/kademlia/kademlia_peer_router.py#L23).

- In Go's implementation [AddrInfo](https://github.com/libp2p/go-libp2p-kad-dht/blob/315504eb159ad07a42fc3a08e17a1b51c84d27b0/routing.go#L498) is the data structure stored for routing in the DHT. Instead of [KadPeerInfo. __repr__](https://github.com/aratz-lasa/py-libp2p/blob/814356383160fb30d376850710d240d9360cea91/libp2p/kademlia/kad_peerinfo.py#L44) that it is currently used. So I have changed to [PeerInfo](https://github.com/libp2p/py-libp2p/blob/00f83a36948fac766be2e2319d59fad86387a87c/libp2p/peer/peerinfo.py#L8), which is Go's [AddrInfo](https://github.com/libp2p/go-libp2p-core/blob/2f75277a8d7b0ee88bd7e9458d054c448f29ec34/peer/addrinfo.go#L11) equivalent, in our implementation. However, in order to store in the current DHT, bytes do not work well (I think that in some RPC bytes are decoded to Strings), strings are the onlye ones working. As a result, I did a temporary quick-fix in [PeerInfo](https://github.com/aratz-lasa/py-libp2p/blob/814356383160fb30d376850710d240d9360cea91/libp2p/peer/peerinfo.py#L9) for decoding and encoding to/from strings.

mhchia commented on libp2p/peer/peerinfo.py:
```diff
@@ -14,6 +15,17 @@ def __init__(self, peer_id: ID, addrs: Sequence[multiaddr.Multiaddr]) -> None:
+    def to_string(self) -> str:
+        return json.dumps([self.peer_id.to_string(), list(map(lambda a: str(a), self.addrs))])
+
+    def __eq__(self, other):
```
```suggestion
    def __eq__(self, other: Any) -> bool:
```

mhchia commented:
Could you also fix the lint error in CI?

> In order to carry out RoutedHost tests, I had to be able to save a TCP addrs for a given Host in the routing DHT. However, find_peer current implementation uses decode_peerinfo, and therefore, create_kad_peerinfo is used. To sum up, currently just UDP addrs can be estored in Kademlia because create_kad_peerinfo loads every peerinfo as if it was a UDP address. So, I have changed find_peer to use a different decoder PeerInfo.info_from_string.

I think it's the implementation difference. We have `Kademlia` protocol in UDP which is specified in the kad paper here. However, [`go-libp2p-kad-dht`](https://github.com/libp2p/go-libp2p-kad-dht) uses the underlying transport, which is possibly TCP or QUIC currently. Cuz our use case(ethereum 2.0) doesn't need `Kademlia` so far, we didn't actively maintain this module and don't have a plan to align it with the go version in the future. For this situation, I'd say it's acceptable that we just have some workaround to make `router` work with `Host`. However, it will still be great if anyone is interested in working on it.

> In Go's implementation AddrInfo is the data structure stored for routing in the DHT. Instead of KadPeerInfo. repr that it is currently used. So I have changed to PeerInfo, which is Go's AddrInfo equivalent, in our implementation. However, in order to store in the current DHT, bytes do not work well (I think that in some RPC bytes are decoded to Strings), strings are the onlye ones working. As a result, I did a temporary quick-fix in PeerInfo for decoding and encoding to/from strings using JSON.

I think it's just fine as long as we make `RoutedHost` work with `Kademlia`, because as mentioned above our current `Kademlia` implementation is not aligned with Go and is not actively maintained.

aratz-lasa commented:
Okay, I already fixed lint error (sorry for so many commits, when I had misconfigured mypy). Then, is it okay to mege? Do I need to add or change something?
I will work in Kademlia DHT little by little. Right now I'm trying to get a draft for the UDP transport.

mhchia commented on libp2p/peer/peerinfo.py:
```diff
@@ -1,19 +1,39 @@
+        )
+
+    @classmethod
+    def info_from_string(cls, info: str) -> "PeerInfo":
```
On second thought, would it be more reasonable to move it to `kademlia_peer_router.py` as a standalone function whose input is `PeerInfo`? Probably something like `peer_info_from_str`. It seems `PeerInfo.info_from_string` is only used in `KadmeliaPeerRouter` and probably not in the rest of the codebase if it is just a workaround?

aratz-lasa commented on libp2p/host/routed_host.py:
```diff
@@ -0,0 +1,40 @@
+            found_peer_info = await self._router.find_peer(peer_info.peer_id)
+            if not found_peer_info:
+                raise ConnectionFailure("Unable to find Peer address")
+            peer_info.addrs = found_peer_info.addrs
```
I thought that too, but in Go's implementation they did, that's why I thought to stick to it. However, I agree it would be better not to. I will change it.

aratz-lasa commented on Pipfile:
```diff
@@ -0,0 +1,12 @@
+[[source]]
```
Don't worry, I already did so. We can stay with setup.py, because there is a lot of debate with Pipfiles...

aratz-lasa commented on libp2p/peer/peerinfo.py:
```diff
@@ -1,19 +1,39 @@
         self.peer_id = peer_id
         self.addrs = list(addrs)
 
+    def to_string(self) -> str:
```
Okay, I think it's a good refactoring.

mhchia commented:
> I will work in Kademlia DHT little by little. Right now I'm trying to get a draft for the UDP transport.

Nice! Would you mind opening an issue discussing your idea in it? And also, do you have a reference to the *UDP transport*? I failed to find UDP transport in `go-libp2p` and `libp2p/specs`. I just want to know your idea and see how we can proceed.

aratz-lasa commented:
I think now It's all done, let me know if it's something I missed. 


> > I will work in Kademlia DHT little by little. Right now I'm trying to get a draft for the UDP transport.
> 
> Nice! Would you mind opening an issue discussing your idea in it? And also, do you have a reference to the _UDP transport_? I failed to find UDP transport in `go-libp2p` and `libp2p/specs`. I just want to know your idea and see how we can proceed.

Okay, I will open an issue and explain my idea in a few hours. I want to develop a small draft before. And I have not searched UDP implementation in Go's repo, but I will do it right now!
